### PR TITLE
Periodically check and remove the taint added by the controller

### PR DESCRIPTION
**Pull request by smshareef**

### CODE REVIEW

File: pkg/hostagent/agent.go [hunk 1]
```diff
@@ -497,22 +478,35 @@ func (agent *HostAgent) runTickers(stopCh <-chan struct{}) {
...
 			}
-			agent.taintRemoved.Store(true)
+			for _, taint := range node.Spec.Taints {
+				if taint.Key == "aci-containers-host/unavailable" {
+					agent.indexMutex.Lock()
+					count := len(agent.completedSyncTypes)
+					agent.indexMutex.Unlock()
+
+					if count == 5 {
+						err := agent.removeTaint(node, "aci-containers-host/unavailable", kubeClient)
```
Comment (akhilamohanan) [inline pkg/hostagent/agent.go:500]: can remove taint only if node is ready?

Comment (smshareef) [reply pkg/hostagent/agent.go:500]: Yes, we can do that.

File: pkg/hostagent/agent.go [hunk 2]
```diff
@@ -497,22 +478,35 @@ func (agent *HostAgent) runTickers(stopCh <-chan struct{}) {
...
+		case <-ticker.C:
+			node, err := kubeClient.CoreV1().Nodes().Get(context.TODO(), os.Getenv("KUBERNETES_NODE_NAME"), metav1.GetOptions{})
 			if err != nil {
-				agent.log.Errorf("Failed to remove taint: %v", err)
 				continue
 			}
-			agent.taintRemoved.Store(true)
+			for _, taint := range node.Spec.Taints {
+				if taint.Key == "aci-containers-host/unavailable" {
+					agent.indexMutex.Lock()
```
Comment (PariyaAshok) [inline pkg/hostagent/agent.go:495]: I think check should based on key as well as effect, We might have same key with different tolerations effect,
Also can we avoid hardcoded key value here?

Comment (smshareef) [reply pkg/hostagent/agent.go]: Thank you for the review Ashok. Made the changes.

File: pkg/hostagent/agent.go (hunk 1 above)
Comment (smshareef) [reply pkg/hostagent/agent.go]: Done Akhila, thanks.

File: pkg/controller/nodes.go [hunk 3]
```diff
@@ -47,6 +47,11 @@ const (
 	tunnelIDIncr = 2
 )
 
+// Name of the taint to add to nodes that are not ready
+const (
+	ACIContainersTaintName string = "aci-containers-host/unavailable"
+)
+
```
Comment (PariyaAshok) [inline pkg/controller/nodes.go:54]: @smshareef  How are we making sure if we are removing taint only added by controller?
With this fix key-name we will remove taint not set by controller right? Can we generate taint-key(key=value:effect) dynamically and set it??

Comment (smshareef) [reply pkg/controller/nodes.go:54]: @PariyaAshok Yes, if the taint with this key and effect NoSchedule is added, then we will remove. We can document that the taint with this key is managed by the ACI-CNI.

If we generate the taint key dynamically, then the same needs to be communicated between the controller and host-agent.

**Review by snaiksat — APPROVED** [review]

**Pull request merged**